src/error.rs

```diff
@@ -21,6 +30,43 @@ pub type DaftResult<T> = std::result::Result<T, DaftError>;
 impl Display for DaftError {
     // `f` is a buffer, and this method must write the formatted string into it
     fn fmt(&self, f: &mut Formatter) -> Result {
-        write!(f, "{self:?}")
+        // Override this method if the error propagated should have custom display for better user-level error messages
+        match self {
+            Self::ExprResolveTypeError {
+                expectation,
+                op_display_name,
+                binary_op_display,
+                fields_to_expr,
+            } => {
+                let simple_field_display = match (binary_op_display, &fields_to_expr[..]) {
+                    (Some(op), [(left_field, _), (right_field, _)]) => {
+                        format!("`{}` {op} `{}`", left_field.name, right_field.name)
+                    }
+                    _ => {
+                        let args: Vec<String> = fields_to_expr
+                            .iter()
+                            .map(|(field, _)| format!("`{}`", field.name))
+                            .collect();
+                        format!("{op_display_name}({})", args.join(", "))
+                    }
+                };
+                let explanation: Vec<String> = fields_to_expr
```

Comment:
no need to make a vec of strings and rewrite it out, unneeded allocation here. You can use `for_each` with `write!` instead instead of map if you have nothing to collect

Reply:
Oic, nice ok

Reply:
Ok just did it in a for loop instead!